File: python/337_HouseRobberIII.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: TreeNode) -> int:
        def helper(root, rob, not_rob):
            if not root: return [0, 0]
            
            left = helper(root.left, rob, not_rob)
            right = helper(root.right, rob, not_rob)
            
            rob = left[1] + right[1] + root.val
            not_rob = max(left[0], left[1]) + max(right[0], right[1])
            
            return [rob, not_rob]
        
        return max(helper(root, 0, 0))

File: swift/337_HouseRobberIII.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func rob(_ root: TreeNode?) -> Int {
        func helper(_ root: TreeNode?) -> [Int] {
            if root == nil { 
                return [0, 0]
            }
            
            var left = helper(root?.left)
            var right = helper(root?.right)
            
            var rob = left[1] + right[1] + root!.val
            var notRob = max(left[0], left[1]) + max(right[0], right[1])
            
            return [rob, notRob]
        }
        
        var result = helper(root)
        return max(result[0], result[1])
    }
}
